crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -1087,6 +1090,21 @@ where
                     .filter(|block_number| *block_number > client.info().finalized_number);
 
                 if let Some(block_number_to_finalize) = maybe_block_number_to_finalize {
+                    {
+                        let mut import_notification = client.every_import_notification_stream();
+
+                        // Drop notification to drop acknowledgement and allow block import to
+                        // proceed
+                        drop(block_importing_notification);
+
+                        while let Some(notification) = import_notification.next().await {
+                            // Wait for importing block to finish importing
+                            if notification.header.number() == &importing_block_number {
```

Comment:
Should it be `notification.header.number() >= &importing_block_number`? otherwise, there can be a race condition where the block is imported before we call `client.every_import_notification_stream()`.

Reply:
We call this from the context of block import, so for practical purposes Substrate right mow will only import one block at a time. But even if not, it doesn't really matter because all we really try to do here is delay finalization, it will work fine even if we don't, just causes unnecessary delays in block propagation as described in https://github.com/autonomys/subspace/issues/3321

Reply:
Okay, I see, the `acknowledgement` in the block notification will avoid such race condition happening. Though, my concern is more about the archiver being stuck in this loop... Perhaps instead of `block_importing_notification_stream`, we should use `every_import_notification_stream` to derive the archiver forward so we don't need to wait for block import to finish.

Reply:
We are using `every_import_notification_stream` though or am I missing something?

Reply:
That is `importing`, not `import` stream. It is already fired for every single block *before* it is imported. We can't replace it with stream that processes blocks *after* import without breaking archiving.